### ISSUE TITLE
a list of use cases

### ISSUE THREAD

**Issue by tim-salabim** [r-spatial/mapedit#19]
A simple list of existing R workflows where we think that mapedit could be useful

- [HistMapR](https://github.com/AGAuffret/HistMapR) -> [function click_sample](https://github.com/AGAuffret/HistMapR/blob/master/R/click_sample.R)

**Comment by mdsumner** [comment]
Editing box geometry models for Atlantis ecosystem models.

Manual correction of initial estimates  for SGAT (and related models) animal track estimation.


Drawing transects or polygons for field programs, e.g. Marine science voyage. (Coupled with extraction from environmental data sets.)

Defining general transect or polygon extraction for localized studies. Nuance via interactivity is critical for things like coastal SST where the exact relationship between the data grain and a hard boundary is locally important. Even simple nuance beyond bbox in longlat is very useful for definition of study areas.

**Comment by tim-salabim** [comment]
Do you have any links to existing solutions/approaches for any of these?

**Comment by mdsumner** [comment]
Yes, sorry was caught out with an unfinished post!   There are so many uses for this for us, I should be trying it out more and making examples. It's a bit unclear to me what you need here, lots of examples to motivate or just a few key specific tasks to solve? Creating/editing a track in space-time is something I would use constantly and give to others to use.

**Comment by fdetsch** [comment]
When downloading higher-level MODIS products using [`runGdal`](https://github.com/MatMatt/MODIS/blob/master/R/runGdal.R) (or [`getHdf`](https://github.com/MatMatt/MODIS/blob/master/R/getHdf.R)) from **MODIS**, users are currently required to 

+ either specify horizontal and vertical tile indices in accordance with the [MODIS Sinusoidal grid](https://modis-land.gsfc.nasa.gov/MODLAND_grid.html) (i.e., look those up manually)
+ or, alternatively, provide a 'Spatial*' or 'Raster*' object from which IDs of intersecting tiles are identified in the background via [`getTile`](https://github.com/MatMatt/MODIS/blob/master/R/getTile.R).

I think it would be great if we could offer a more flexible, interactive way of how to select tiles, which especially involves 

+ selecting single tiles by a) displaying the global MODIS grid on a **leaflet** map, b) clicking relevant polygons and c) returning **sf** objects back to R
+ as well as allowing the user to draw polygons from which to derive corresponding tile IDs under the hood.

**Comment by tim-salabim** [comment]
Great idea! Can you provide a sample tile grid data set (preferably as sf)?

**Comment by edzer** [comment]
Cool idea; I've been thinking about the idea of handling tile collections quite a bit, lately. 

A set of tiles for sentinel 2 is loaded by:
```r
tiles = st_read("https://sentinel.esa.int/documents/247904/1955685/S2A_OPER_GIP_TILPAR_MPC__20151209T095117_V20150622T000000_21000101T000000_B00.kml")
Reading layer `Features' from data source `https://sentinel.esa.int/documents/247904/1955685/S2A_OPER_GIP_TILPAR_MPC__20151209T095117_V20150622T000000_21000101T000000_B00.kml' using driver `KML'
converted into: GEOMETRYCOLLECTION
Simple feature collection with 56686 features and 2 fields
geometry type:  GEOMETRYCOLLECTION
dimension:      XYZ
bbox:           xmin: -180 ymin: -83.83595 xmax: 180 ymax: 84.64428
epsg (SRID):    4326
proj4string:    +proj=longlat +datum=WGS84 +no_defs
plot(tiles[1], xlim = c(0,10), ylim = c(0,10))
```

![tiles](https://cloud.githubusercontent.com/assets/520851/26055012/5163ae42-396f-11e7-8aba-61dc384b5602.png)

These are UTM tiles, laid out in longlat; it would be great if we could handle (subsets of) these in R. My current idea is to do this the way [PostGIS raster](https://postgis.net/docs/RT_reference.html) does this: in a table, with each record containing, or referring, to a single tile, and its metadata (e.g. outline as an `sfg`).

**Comment by mdsumner** [comment]
I was just messing around with the sinu grid, getting their largest tiles as per the diagram:  https://gist.github.com/mdsumner/19b36c829ec1cbedabff1bf619b4ec18

It would be interesting to abstract that to zoom as well, but I think easiest would be to store empty raster layers the grid contstructs, and use the cell logic to generate them. 

I see the long instructions for manually creating a grid for the poles with the .hdr files too, happy to help iron that out if needed.

**Comment by fdetsch** [comment]
An EPSG:4326 version of the MODIS tile grid is shipped with the package. It originates from before my time, so I am not quite sure where Matteo took it from. But anyway, it can be loaded into R using

```
library(sf)
library(mapview)

if (!require(MODIS)) 
  devtools::install_github("MatMatt/MODIS", ref = "develop")

shp <- system.file("external", "modis_latlonWGS84_grid_world.shp", package = "MODIS")
grd <- st_read(shp)

m <- mapview(grd)
m
```

![modis_grid](http://i.imgur.com/okAi9YF.png)

**Comment by mdsumner** [comment]
Thanks @fdetsch  - that is helpful -  you can make more sense of the plot when you do this: 

```R
plot(st_transform(grd, st_crs("+proj=sinu"))
```

![image](https://cloud.githubusercontent.com/assets/4107631/26056094/2b2d33b8-39b7-11e7-9a54-c161a82fa0e2.png)

and then to get it into mapview with sufficient density of vertices along edges, this works but there's only so far we can push the sense of mapping in Web Mercator here: 

```R
stz <- st_segmentize(st_set_crs(grd, NA), dfMaxLength=.05)
mapview(st_set_crs(stz, 4326))

```

(You can't segmentize with the CRS set because sf will follow great circles, so distance calc now has to occur in native space for non-orthodromes). 

Note that this regular tiling in the sinusoidal projection is apposite to the integrized sinusoidal grid used by "L3bin" products. I don't expect that will come up in this context, but it's a terminology-clash to watch for.

**Comment by fdetsch** [comment]
Thanks for pointing this out @mdsumner, very useful stuff!

**Comment by tim-salabim** [comment]
I had a quick play w the MODIS example and this seems to work (Note that we need to set a class `character` grouping in order to get this working)

```r
grd$cat = as.character(grd$cat)

m = leaflet() %>% 
    addTiles() %>% 
    addPolygons(data = grd, weight = 1, group = ~cat)

select_map(m,
           style_false = list(weight = 1),
           style_true = list(weight = 4))
```

**Comment by tim-salabim** [comment]
With 4dfd8d94edfe5378e4b53c36003b63c04ee5a439 we can now do

```r
library(mapview)
library(sf)

subs = selectFeatures(breweries, style_true = list(fillColor = "red"))
## or
subs = selectFeatures(trails, style_true = list(fillColor = "red"))
## or
subs = selectFeatures(franconia, style_true = list(fillColor = "red"))
## which is similar to 
subs = selectFeatures(grd, style_true = list(fillColor = "red")) # taking grd from the example above
```
All of these will return a `sf` object, hence
```r
mapview(subs)
```
should work for all of the above.

This is, however, just a rough first implementation as there are still many assumptions and it's not at all flexible as of now. Still, this may be one road we could take. 
Feedback is more than welcome! Especially from @timelyportfolio

**Comment by timelyportfolio** [comment]
Can't wait to check it out.  Thanks everyone!

**Comment by fdetsch** [comment]
Looks great, thanks everyone. I'll create a new **MODIS** branch at [https://github.com/MatMatt/MODIS](https://github.com/MatMatt/MODIS) and add this example to `getTile()`. This will definitely find its way into our talk at this year's useR! (:

**Comment by timelyportfolio** [comment]
@tim-salabim, are you happy with the `style_true` argument?  I have not found any better pattern, and I love the flexibility, but this does require some knowledge of `CSS`.

**Comment by tim-salabim** [comment]
@timelyportfolio yes, people should be familiar with the CSS parameters from the likes of leaflet::pathOptions and especially leaflet::highlightOptions. List semantic is also fine. We can provide a suitable function to build the list at any stage without problem

**Comment by timelyportfolio** [comment]
@fdetsch, any feedback on @tim-salabim's very nice first approach?  Thanks again for this very helpful use case.

**Comment by timelyportfolio** [comment]
Here similar to above example with `st_voronoi`.

```
# example ?st_voronoi
x = st_multipoint(matrix(runif(30),,2))
box = st_polygon(list(rbind(c(0,0),c(1,0),c(1,1),c(0,1),c(0,0))))
if (sf_extSoftVersion()["GEOS"] >= "3.5.0") {
  v = st_sfc(st_voronoi(x, st_sfc(box)))
  plot(v, col = 0, border = 1, axes = TRUE)
  plot(box, add = TRUE, col = 0, border = 1) # a larger box is returned, as documented
  plot(x, add = TRUE, col = 'red', cex=2, pch=16)
  plot(st_intersection(st_cast(v), box)) # clip to smaller box
  plot(x, add = TRUE, col = 'red', cex=2, pch=16)
}
# select voronoi polygons
voronoi_selected <- selectFeatures(
  st_sf(st_cast(v)),
  styleFalse = list(fillColor="transparent", weight=1),
  styleTrue = list(fillColor="pink", weight=3)
)
plot(voronoi_selected)
```

**Comment by mdsumner** [comment]
I want to share this while it's fresh, it's not something I tend to think I need: 

Straight mapview(any-thing or just tiles) and  digitizing points for use later is a pretty simple but open opportunity I think. I just opened up Manifold for 5 minutes with a colleague to collect several dozen points from an old krill figure as longlats - and they were amazed - I forget many people just don't have the facilities to do that. I could have use mapedit but 

* the rough GeoTIFF had to be exported and read into R
* plotting in R would wrap it into WebMercator (from polar LAEA) look funny, or
* proper plotting would require complex set up with a polar leaflet projection and rare tile providers
* drawing points would require clicking "add marker" every time
* the result would be sf rather than a dataframe of long-lat

It might be worth presenting a simple-as-possible example of doing just that. It's totally needs-driven and exactly fit for purpose on the day it's needed, though usually a feature we tend to look past to more complex things. 

In this case I'd probably wrap the whole set up into one function, like "pdigitize()" that simple returned a data frame of longitude latitude. The UI would need just a "add multi point" button, enabled by default. 

Also realted, I sometimes do a lot of rough affine-based georeferencing of maps sometimes, and I always use Manifold or my own locator-driven DIY method in R/raster, and rarely something crafted with VRT and GDAL. I'll try work out an example for mapedit, it requires side-by-side maps one in plain view and one in the native projection .

**Comment by fdetsch** [comment]
@timelyportfolio, I've opened a new **MODIS** branch called [mapedit](https://github.com/MatMatt/MODIS/tree/mapedit) and implemented Tim's approach in [`getTile()`](https://github.com/MatMatt/MODIS/blob/mapedit/R/getTile.R). It works every now and then, but then again, it fails at least as often. But see for yourselves: 

![getTile](http://i.imgur.com/vVl1G0z.gif)

**Comment by timelyportfolio** [comment]
thanks @fdetsch for the screenshot.  I will explore and debug.  Sorry for the trouble.

**Comment by fdetsch** [comment]
Sounds good, thank you! In addition, in order to expand interactive tile selection a little bit, it would also be great if we could have `editMap()` return an object of class 'sf' based on which we would then be able to identify intersecting MODIS tiles.

**Comment by timelyportfolio** [comment]
@fdetsch, `editMap()` returns `sf` as of https://github.com/r-spatial/mapedit/pull/18.  However, I did not realize when I wrote the `geojson -> sf` conversion that there was already a way in `sfr` to convert (see [issue comment](https://github.com/edzer/sfr/issues/185#issuecomment-304093929)).  This might offer a more robust solution, but I have not had the opportunity to test.

**Comment by timelyportfolio** [comment]
@fdetsch, @tim-salabim promoted `selectFeatures` to its own issue #24.

**Comment by tim-salabim** [comment]
@mdsumner I moved your 'fresh' thouhgts to [#25](https://github.com/r-spatial/mapedit/issues/25)

**Comment by mdsumner** [comment]
With a colleague we just put together a workflow that changed a very manual tiling and image-exploration task into a straightforward set of steps, useable for teaching and real work. This started as a question about how to tile an image with a buffer, and 45 minutes later we had a working prototype ready for some final tweaks. :)

Here is all the code. The list of files is not reproducible, it's just a set of GeoTIFFs on a local drive (in UTM of various zones). 

The steps are

1. load an image and produce a tile index of it at given grain (500m)
2. set up a sampled subset of the tiles to be intensively searched
3. set up the mapview object with the right setting (the image scaled appropriately, and the tile region overlad)
4. user interaction, the user interacts with one of the sampled tiles to search the image for features (seals in this case), they can see the tile border as well as a 50m buffer of the larger image outside of that tile, in order to avoid edge-effects and possibly missing counting a feature
5. follow up with the drawn output, plot independently in another context

Thanks!  This is so awesome. 


```R
library(mapedit)
library(mapview)
library(raster)
library(spex)
img.list <- dir(pattern=".TIF")
p1 <- raster(img.list[1])

## 1. -----------------------------------------------------------------------------
## grid of tiles at 500m grain
grain <- 500
buf <- 50
tiles <- qm_rasterToPolygons_sp(raster(buffer_extent(extent(p1), grain), 
                                       crs = projection(p1), res = grain))

## 2. -----------------------------------------------------------------------------
## we treat this as a sampling exercise, a subset of the tiles
## wll be searched intensively to illustrate issues of study and sampling design
nsamples <- 10
tiles$sample <- FALSE
tiles$sample[sample(nrow(tiles), nsamples)] <- TRUE
#plot(tiles, col = factor(tiles$sample + 1))

# 3. -----------------------------------------------------------------------------
## presentation of the map
ss <- which(tiles$sample)
i <- 1  ## of sum(tiles$sample)
tile_image <-crop(p1, extent(tiles[ss[i], ]) + buf)
col_map <- grey(seq(0, 1, length = 256))
val_map <- seq(cellStats(p1, min), cellStats(p1, max))
map1 <- mapview(tile_image, maxpixels = ncell(tile_image), 
                col.regions = col_map, 
                values = val_map) + mapview(as(tiles[ss[i], ], "SpatialLines"))

# 4. user interaction -----------------------------------------------------------------------------
## the editing session, user sees the "focus tile" as well as the image within
## that tile, but also an extra border of image to width "buf"
## the user's task is to "find seals" and click on where they are
em <- map1 %>% editMap()

# 5. follow up -----------------------------------------------------------------------------
## remember we get the drawn result in 4326 so need to transform
tile_i_points <- sf::st_transform(em$drawn, projection(p1))
plot(tile_image)
plot(tile_i_points, add = TRUE, col = "hotpink")
```
